From 7a39d433f6490618958f8ab295476914a2d44c4d Mon Sep 17 00:00:00 2001
From: Long Dang <long.dang.aj@rvc.renesas.com>
Date: Thu, 18 Aug 2016 12:07:27 +0700
Subject: [PATCH 41/42] gst145-omxvideodec: Add a support for the page
 alignment for NV12 planes

Porting from a patch in Linux 3.10 environment made by Kazunori Kobayashi <kkobayas@igel.co.jp>

The feature is optional and can be enabled by running ./configure
with "--nv12-page-alignment".

Signed-off-by: Long Dang <long.dang.aj@rvc.renesas.com>
---
 configure.ac         | 12 +++++++++
 omx/gstomxvideodec.c | 73 ++++++++++++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 85 insertions(+)

diff --git a/configure.ac b/configure.ac
index e66efd1..44f4890 100644
--- a/configure.ac
+++ b/configure.ac
@@ -277,6 +277,18 @@ AC_CHECK_LIB([mmngrbuf], [mmngr_export_start_in_user],
              ])
 fi
 
+dnl check page alignment option for NV12 planes
+AC_ARG_ENABLE([nv12-page-alignment],
+              [AS_HELP_STRING([--enable-nv12-page-alignment],
+                              [Enable the page alignment feature for NV12 planes])],
+              [],
+              [enable_nv12_page_alignment=no]
+)
+if test "x${enable_nv12_page_alignment}" = "xyes"; then
+  AC_DEFINE(ENABLE_NV12_PAGE_ALIGN, 1,
+            [Define if the page alignment feature for NV12 planes is enabled])
+fi
+
 dnl *** set variables based on configure arguments ***
 
 dnl set license and copyright notice
diff --git a/omx/gstomxvideodec.c b/omx/gstomxvideodec.c
index 97e66ec..bca11fd 100644
--- a/omx/gstomxvideodec.c
+++ b/omx/gstomxvideodec.c
@@ -48,6 +48,7 @@
 #endif
 
 #include <string.h>
+#include <unistd.h>             /* getpagesize() */
 
 #include "gstomxbufferpool.h"
 #include "gstomxvideo.h"
@@ -2073,6 +2074,59 @@ gst_omx_video_dec_negotiate (GstOMXVideoDec * self)
   return (err == OMX_ErrorNone);
 }
 
+#ifdef ENABLE_NV12_PAGE_ALIGN
+static gint
+get_uv_offset_aligned_to_page (gint page_size, gint stride, gint height)
+{
+  gint a, b, r;
+  gint lcm;
+
+  /*
+   * The following implementation uses the Euclidean Algorithm to obtain
+   * the least common multiple of stride and page size.
+   */
+
+  /* nStride is set to width, to achieve 4K aligned by adjusting
+     the nSliceHeight. */
+  /* (1) Calculate the GCD of stride and alignment */
+  b = stride;
+  a = page_size;
+  while ((r = a % b) != 0) {
+    a = b;
+    b = r;
+  }
+
+  /* (2) Calculate the LCM of stride and alignment */
+  lcm = stride * page_size / b;
+
+  /* (3) Calculate the offset of UV plane */
+  return (((stride * height) / lcm) + 1) * lcm;
+}
+
+static gboolean
+gst_omx_video_dec_align_uv_offset_to_page (GstOMXVideoDec * self,
+    OMX_PARAM_PORTDEFINITIONTYPE * out_port_def, gint page_size, gint stride,
+    gint height)
+{
+  gint uv_offset;
+
+  uv_offset = get_uv_offset_aligned_to_page (page_size, stride, height);
+
+  out_port_def->format.video.nStride = stride;
+  out_port_def->format.video.nSliceHeight = uv_offset / stride;
+
+  GST_DEBUG_OBJECT (self,
+      "Set nSliceHeight to %u for aligning the UV plane offset to the page size",
+      (guint) out_port_def->format.video.nSliceHeight);
+
+  if (gst_omx_port_update_port_definition (self->dec_out_port,
+          out_port_def) != OMX_ErrorNone)
+    return FALSE;
+
+  return TRUE;
+}
+#endif /* ENABLE_NV12_PAGE_ALIGN */
+
 static gboolean
 gst_omx_video_dec_set_format (GstVideoDecoder * decoder,
     GstVideoCodecState * state)
@@ -2083,6 +2137,10 @@ gst_omx_video_dec_set_format (GstVideoDecoder * decoder,
   gboolean is_format_change = FALSE;
   gboolean needs_disable = FALSE;
   OMX_PARAM_PORTDEFINITIONTYPE port_def;
+#ifdef ENABLE_NV12_PAGE_ALIGN
+  OMX_PARAM_PORTDEFINITIONTYPE out_port_def;
+  gint page_size;
+#endif
 
   self = GST_OMX_VIDEO_DEC (decoder);
   klass = GST_OMX_VIDEO_DEC_GET_CLASS (decoder);
@@ -2269,6 +2327,21 @@ gst_omx_video_dec_set_format (GstVideoDecoder * decoder,
     if (!gst_omx_video_dec_negotiate (self))
       GST_LOG_OBJECT (self, "Negotiation failed, will get output format later");
 
+#ifdef ENABLE_NV12_PAGE_ALIGN
+    gst_omx_port_get_port_definition (self->dec_out_port, &out_port_def);
+
+    page_size = getpagesize ();
+    if (out_port_def.format.video.eColorFormat ==
+        OMX_COLOR_FormatYUV420SemiPlanar &&
+        (info->width * info->height) & (page_size - 1))
+      if (!gst_omx_video_dec_align_uv_offset_to_page (self, &out_port_def,
+              page_size, info->width, info->height)) {
+        GST_ERROR_OBJECT (self,
+            "Failed to align the uv offset of the NV12 plane to the page size");
+        return FALSE;
+      }
+#endif
+
     if (!(klass->cdata.hacks & GST_OMX_HACK_NO_DISABLE_OUTPORT)) {
       /* Disable output port */
       if (gst_omx_port_set_enabled (self->dec_out_port, FALSE) != OMX_ErrorNone)
-- 
1.9.1

